PowerEditor/src/WinControls/ProjectPanel/DirectoryWatcher.cpp

```diff
@@ -0,0 +1,389 @@
+	: _hWnd(hWnd)
+	, _hThread(NULL)
+	, _hRunningEvent(NULL)
+	, _hStopEvent(NULL)
+	, _hUpdateEvent(NULL)
+	, _running(false)
+	, _updateFrequencyMs(updateFrequencyMs)
+	, _changeOccurred(false)
+	, _hideEmptyDirs(hideEmptyDirs)
+	, _checkEmptyDirsCount(0)
+	, _checkEmptyDirsEvery(checkEmptyDirsEvery)
+	, _manualMode(manualMode)
+{
+}
+
+DirectoryWatcher::~DirectoryWatcher()
+{
+	Yuni::MutexLocker lock(_lock);
+	stopThread();
+	removeAllDirs();
+}
+
+void DirectoryWatcher::addOrChangeDir(const generic_string& path, HTREEITEM treeItem, const std::vector<generic_string>& filters)
+{
+	Yuni::MutexLocker lock(_lock);
+	
+	// remove a probably previously existing item
+	// to easily change the filters.
+	_dirItemsToRemove.insert(treeItem);
+
+	// each newly added item gets a forced update message.
+	_forcedUpdateToAdd.insert(treeItem);
+	_dirItemsToAdd.emplace_back(new InsertStruct(path,treeItem,filters));
+}
+
+void DirectoryWatcher::removeDir(HTREEITEM treeItem)
+{
+	Yuni::MutexLocker lock(_lock);
+	_dirItemsToRemove.insert(treeItem);
+}
+
+void DirectoryWatcher::removeAllDirs()
+{
+	Yuni::MutexLocker lock(_lock);
+	for (auto it=_watchdirs.begin(); it != _watchdirs.end(); ++it)
+		delete *it;
+	_watchdirs.clear();
+
+	_dirItems.clear();
+	_dirItemReferenceCount.clear();
+	_forcedUpdate.clear();
+
+	for (auto it=_dirItemsToAdd.begin(); it != _dirItemsToAdd.end(); ++it)
+		delete *it;
+	_dirItemsToAdd.clear();
+	_dirItemsToRemove.clear();
+	_forcedUpdateToAdd.clear();
+	
+}
+
+void DirectoryWatcher::setManualMode(bool val)
+{
+
+	Yuni::MutexLocker lock(_lock);
+	if (_manualMode != val)
+	{
+		_manualMode = val;
+		// wake up thread if manual mode has been switched off
+		if (!_manualMode)
+			update();
+	}
+}
+
+void DirectoryWatcher::forceUpdate(HTREEITEM hItem)
+{
+	Yuni::MutexLocker lock(_lock);
+	assert( _dirItems.find(hItem) != _dirItems.end());
+	if( _dirItems.find(hItem) != _dirItems.end())
+	{
+		_forcedUpdateToAdd.insert(hItem);
+		update();
+	}
+}
+
+void DirectoryWatcher::forceUpdateAll()
+{
+	Yuni::MutexLocker lock(_lock);
+	for (auto it=_dirItems.begin(); it !=_dirItems.end(); ++it)
+	{
+		_forcedUpdateToAdd.insert(it->first);
+	}
+	update();
+}
+
+void DirectoryWatcher::update()
+{
+	::SetEvent(_hUpdateEvent);
+}
+
+void DirectoryWatcher::startThread()
+{
+
+	try {
+		Yuni::MutexLocker lock(_lock);
+		if (_running)
+			return;
+
+		_hRunningEvent = ::CreateEvent(NULL, TRUE, FALSE, NULL);
+		_hStopEvent = ::CreateEvent(NULL, TRUE, FALSE, NULL);
+		_hUpdateEvent = ::CreateEvent(NULL, TRUE, FALSE, NULL);
+		if (!_hStopEvent || !_hRunningEvent || !_hUpdateEvent)
+			throw std::runtime_error("Could not create DirectoryWatcher events");
+
+		_hThread = ::CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)threadFunc, (LPVOID)this, 0, NULL);
+		if (!_hThread)
+			throw std::runtime_error("Could not create DirectoryWatcher thread");
+
+		::WaitForSingleObject(_hRunningEvent, INFINITE);
+	}
+	catch (...)
```

Comment:
Why the `catch (...)`?? You're only ever `throw`ing a `std::runtime_error`. Even worse, given `Notepad++`'s compilation options, this will also catch Access Violations!